utils/swift_build_support/swift_build_support/products/earlyswiftdriver.py

```diff
@@ -0,0 +1,133 @@
+                           '(Try `--skip-early-swift-driver`)'
+                print('-- Warning: {}', warn_msg)
+                return False
+            else:
+                return True
+        return False
+
+    @classmethod
+    def get_dependencies(cls):
+        return []
+
+    def should_clean(self, host_target):
+        return self.args.clean_early_swift_driver
+
+    def clean(self, host_target):
+        run_build_script_helper('clean', host_target, self, self.args)
+
+    def build(self, host_target):
+        run_build_script_helper('build', host_target, self, self.args)
+
+    def should_test(self, host_target):
+        # EarlySwiftDriver is tested against the compiler's lit test
+        # suite driver subset, as a standalone test CMake target when
+        # `swift` is built. We do not run the driver's own tests here.
+        return False
+
+    def test(self, host_target):
+        run_build_script_helper('test', host_target, self, self.args)
+
+    def should_install(self, host_target):
+        # This product is for the swift-driver used with the build-directory compiler.
+        # If a toolchain install is required, please use the SwiftDriver (no 'Early')
+        # product with `--swift-driver --install-swift-driver`.
+        return False
+
+    def install(self, host_target):
+        run_build_script_helper('install', host_target, self, self.args)
+
+
+def run_build_script_helper(action, host_target, product, args):
+    build_root = os.path.dirname(product.build_dir)
+    script_path = os.path.join(
+        product.source_dir, 'Utilities', 'build-script-helper.py')
+
+    # Building with the host toolchain for use with a local compiler build,
+    # use the toolchain which is supplying the `swiftc`.
+    swiftc_path = os.path.abspath(product.toolchain.swiftc)
+    toolchain_path = os.path.dirname(os.path.dirname(swiftc_path))
+
+    # Pass Dispatch directory down if we built it
+    dispatch_build_dir = os.path.join(
+        build_root, '%s-%s' % ('libdispatch', host_target))
+
+    # Pass Foundation directory down if we built it
+    foundation_build_dir = os.path.join(
+        build_root, '%s-%s' % ('foundation', host_target))
+
+    # Pass the swift lit tests if we're testing and the Swift tests were built
+    swift_build_dir = os.path.join(
+        build_root, 'swift-{}'.format(host_target))
+    lit_test_dir = os.path.join(
+        swift_build_dir, 'test-{}'.format(host_target))
+
+    is_release = product.is_release()
+    configuration = 'release' if is_release else 'debug'
+    helper_cmd = [
+        script_path,
+        action,
+        '--package-path', product.source_dir,
+        '--build-path', product.build_dir,
+        '--configuration', configuration,
+        '--toolchain', toolchain_path,
+        '--ninja-bin', product.toolchain.ninja,
+        '--cmake-bin', product.toolchain.cmake,
+        '--local_compiler_build'
+    ]
+
+    if os.path.exists(dispatch_build_dir):
```

Comment:
Question. Should we error here if we do not have these build directories? How does the failure of these conditions manifest?

Reply:
As far as I remember, this is platform specific. On Linux, `build-script` propagates an explicit Foundation build dir to the products, which is required. If that is missing, then the driver's `build-script-helper.py` will fail. On Darwin, it doesn't matter because we expect Foundation to just be found on the host system. 

In any case, I think if there is an issue with this, we should fix it separately from this change, for both the (driver) build products.

Reply:
Ok. SGTM. I was just thinking about the failure modes on Linux.